teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,833 @@
+      [using them made a Task less resuable](https://github.com/tektoncd/pipeline/blob/main/docs/resources.md#why-arent-pipelineresources-in-beta) -
+      this proposal introduces an alternative that does not involve making a Task less reusable (especially
+      once coupled with [TEP-0044](https://github.com/tektoncd/community/blob/main/teps/0044-decouple-task-composition-from-scheduling.md))
+* [Simplicity](https://github.com/tektoncd/community/blob/main/design-principles.md#simplicity)
+    * Pro: Support for emitting object results builds on [TEP-0076](https://github.com/tektoncd/community/pull/477)
+    * Pro: This proposal reuses the existing array or string concept for params
+    * Pro: This proposal continues [the precedent of using JSONPath syntax in variable replacement](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-561476075)
+* [Flexibility](https://github.com/tektoncd/community/blob/main/design-principles.md#flexibility)
+    * Pro: Improves support for structured interfaces which tools can rely on
+    * Con: Although there is a precedent for including JSONPath syntax, this is a step toward including more hard coded
+      expression syntax in the Pipelines API (without the ability to choose other language options)
+    * Con: We're also introducing and committing to json schema! Seems worth it for what we get tho
+* [Conformance](https://github.com/tektoncd/community/blob/main/design-principles.md#conformance)
+    * Supporting this syntax would be part of the conformance surface; the json schema syntax is a bit verbose for
+      simple cases (but paves hte way for the more complex cases we can support later, including letting Tasks
+      express how to validate their own parameters)
+
+## Drawbacks
+
+[See also TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477) for more drawbacks.
+
+* In the current form, this sets a precedent for pulling in more json schema support (or whatever format we choose
+  for expressing the schema)
+* This improves the ability to make assertions about Task interfaces but doesn't (explicitly) include worksapces at all.
+  For example, say you had a “git” param for a pipeline containing a commitish and a url. This is likely going to be
+  used to clone from git into a workspace, and there will be some guarantees that can be made about the data that ends
+  up in the workspace (e.g. the presence of a .git directory). But there is no way of expressing the link between the
+  workspace and the params.
+
+## Alternatives
+
+## Alternative #1: Introduce minimal shorthand for json schema
+
+If we wanted to make our adoption of the json schema syntax a less verbose, we could make a couple of changes:
+
+- We could imply `type:string` for keys by default (especially since this is the only initially supported type)
+- We could remove the top level schema section and embed the details in the param/result definition.
```

Comment:
TODO: ive just realized that having a top level schema section is actually something we added and not a json schema specific thing so i want to update the proposal above to not use the top level schema section, and add it as an alternative instead